### PR TITLE
Feature/movie details screen

### PR DESCRIPTION
**Pull request by evakiraly**
Screenshot 2021-11-08 at 09 29 51
Screenshot 2021-11-08 at 09 29 15

### CODE REVIEW

File: EnMovie/Modules/Details/CustomDetailsView.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+   
+    // MARK: - UI
+
+    func initUI() {
+        detailsStackView.addArrangedSubview(detailLabel(movieYear ?? "-"))
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/CustomDetailsView.swift:69]: remove the moviewYear ?? - and have a separate lazy var label for everything and set the text separately

File: EnMovie/Modules/Details/CustomDetailsView.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        detailsStackView.addArrangedSubview(simpleSeparatorView())
+        detailsStackView.addArrangedSubview(detailLabel(runtime ?? "-"))
+        detailsStackView.addArrangedSubview(simpleSeparatorView())
+        detailsStackView.addArrangedSubview(detailLabel(categories ?? "-"))
+        
+        self.addSubview(detailsStackView)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/CustomDetailsView.swift:75]: self is no needed (respect the guidelines)

File: EnMovie/Modules/Details/CustomDetailsView.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    
+    func initConstraints() {
+        detailsStackView.translatesAutoresizingMaskIntoConstraints = false
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/CustomDetailsView.swift:79]: you can set this in lazy var

File: EnMovie/Modules/Details/CustomRatingView.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
+//
+//  DetailsCustomView.swift
+//  EnMovie
+//
+//  Created by Eva Kiraly on 28.10.2021.
+//
+
+import Foundation
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/CustomRatingView.swift:8]: similar as in another file

File: EnMovie/Modules/Details/CustomRatingView.swift [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    
+    func initConstraints() {
+        let constraints = [
+            ratingStackView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
+            ratingStackView.leadingAnchor.constraint(equalTo: self.leadingAnchor),
+            ratingStackView.topAnchor.constraint(equalTo: self.topAnchor),
+            ratingStackView.bottomAnchor.constraint(equalTo: self.bottomAnchor)
+        ]
+        NSLayoutConstraint.activate(constraints)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/CustomRatingView.swift:81]: all the comments from the previous view applies here

File: EnMovie/Modules/Details/DetailsPresenter.swift [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+// MARK: - DetailsInteractorOutput
+
+extension DetailsPresenter: DetailsInteractorOutput {
+    func didGetMovieDetails(_ movie: MovieDetailsDTO) {
+        view.didGetResult(result: movie)
+        view.hideIndicator()
+    }
+    
+    func didGetError(_ error: Error) {
+        print(error.localizedDescription)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsPresenter.swift:42]: should we show and error alert?

File: EnMovie/Modules/Details/DetailsRouter.swift [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+//  DetailsRouter.swift
+//  EnMovie
+//
+//  Created by Eva Kiraly on 27.10.2021.
+//
+
+import UIKit
+
+protocol DetailsRouterInput {
+}
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsRouter.swift:11]: you can keep it on one line

File: EnMovie/Modules/Details/DetailsRouter.swift [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+protocol DetailsRouterOutput: AnyObject {
+}
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsRouter.swift:14]: similar

File: EnMovie/Modules/Details/DetailsRouter.swift [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    init(presentingViewController: UIViewController) {
+        self.presentingViewController = presentingViewController
+    }
+}
+
+// MARK: - DetailsRouterInput
+
+extension DetailsRouter: DetailsRouterInput {
+}
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsRouter.swift:28]: one line

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+//  DetailsViewController.swift
+//  EnMovie
+//
+//  Created by Eva Kiraly on 27.10.2021.
+//
+
+import UIKit
+
+protocol DetailsViewInput: AnyObject {
+    func didGetResult(result: MovieDetailsDTO)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:11]: maybe have an alert for error case?

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    @IBOutlet private var storylineLabel: UILabel!
+    @IBOutlet private var storylineDescriptionLabel: UILabel!
+    @IBOutlet private var containerView: UIView!
+    private var scrollView = UIScrollView()
+    private var imageContainerView = UIView()
+    private var verticalGradient: CAGradientLayer!
+    private var horizontalGradient: CAGradientLayer!
+    private var provider: MovieDetailsProvider!
+        
+    private var ratingView: CustomRatingView = {
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:41]: lazy var :D

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        let view = CustomRatingView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
+    }()
+
+    private var basicMovieInfosView = {(movieYear: String?, runtime: String?, categories: String?) -> CustomDetailsView in
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:47]: lazy var and have a public method for setting year, runtime, categories

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+final class DetailsViewController: UIViewController {
+    var output: DetailsViewOutput?
+    var actualMovieId: Int?
+
+    // MARK: - Private properties
+    
+    private var movie: MovieDetailsDTO?
+
+    @IBOutlet var indicatorView: UIActivityIndicatorView!
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:28]: private?

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    func viewIsReady(movieId: Int)
+}
+
+final class DetailsViewController: UIViewController {
+    var output: DetailsViewOutput?
+    var actualMovieId: Int?
+
+    // MARK: - Private properties
+    
+    private var movie: MovieDetailsDTO?
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:26]: move it with other private vars

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+protocol DetailsViewOutput {
+    func viewIsReady(movieId: Int)
+}
+
+final class DetailsViewController: UIViewController {
+    var output: DetailsViewOutput?
+    var actualMovieId: Int?
+
+    // MARK: - Private properties
+
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:25]: new line?

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    private var basicMovieInfosView = {(movieYear: String?, runtime: String?, categories: String?) -> CustomDetailsView in
+        let view = CustomDetailsView()
+        view.configure(year: movieYear, runtime: runtime, categories: categories)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
+    }
+
+    // MARK: - Lifecycle
+
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:55]: new line not needed

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+        scrollView.isScrollEnabled = true
+        scrollView.alwaysBounceVertical = true
+        scrollView.isUserInteractionEnabled = true
+        scrollView.delaysContentTouches = true
+        
+        stackView.axis = .vertical
+        stackView.distribution = .equalSpacing
+        stackView.spacing = 8.0
+        
+        containerView.addSubview(imageContainerView)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:87]: do we need addsubview? is not already included in storyboard?

File: EnMovie/Modules/Details/DetailsViewController.swift [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+        containerView.addSubview(stackView)
+        
+        scrollView.addSubview(containerView)
+        
+        view.backgroundColor = .customBackgroundColor
+        view.addSubview(scrollView)
+    }
+    
+    func initConstraints() {
+        containerView.translatesAutoresizingMaskIntoConstraints = false
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/DetailsViewController.swift:101]: you can add constraints in storyboard? if not you can select translateAutoresizingMask = false from there

File: EnMovie/Modules/Details/Entity/Movie.swift [hunk 19]
```diff
@@ -0,0 +1,22 @@
+//
+//  Movie.swift
+//  EnMovie
+//
+//  Created by Eva Kiraly on 02.11.2021.
+//
+
+import Foundation
+import UIKit
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/Entity/Movie.swift:9]: do you need UIKit? it's a model file

File: EnMovie/Modules/Details/MovieDetailsProvider.swift [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    init(session: URLSession) {
+        self.session = session
+    }
+    
+    func getMovieDetails(id: Int, completion: @escaping (Result<MovieDetailsDTO, Error>) -> Void) {
+
+        guard let urlString = URL(string: "https://api.themoviedb.org/3/movie/\(id)?api_key=\(apiKey)&language=en-US&page=1") else { return }
+        
+        let task = session.dataTask(with: urlString) { data, response, error in
+            if let error = error {
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Details/MovieDetailsProvider.swift:27]: use guard, example: guard let data = data and try decoded, an with else return the completion Error (you can change it to be optional in escaping)

File: EnMovie/Modules/Home/Presenter/HomePresenter.swift [hunk 21]
```diff
@@ -29,9 +29,14 @@ final class HomePresenter {
...
 
 extension HomePresenter: HomeViewOutput {
+    func cellWasSelected(with movieId: Int) {
+        router.presentDetailsScreen(with: movieId)
+    }
+    
     func viewIsReady() {
         interactor.getMoviesData(parameterType: .upcoming)
     }
+
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Home/Presenter/HomePresenter.swift:39]: new line not needed

File: EnMovie/Modules/Home/View/Reusable Views/MoviesTableViewCell.swift [hunk 22]
```diff
@@ -7,13 +7,20 @@
...
+
 class MoviesTableViewCell: UITableViewCell {
     
     private var upcomingMoviesArray: [Movies] = []
     private let collectionViewEdgeInsets = UIEdgeInsets(top: 5, left: 15, bottom: 5, right: 15)
     @IBOutlet private weak var collectionView: UICollectionView!
     @IBOutlet private weak var collectionViewHeightConstraint: NSLayoutConstraint!
     
+    weak var movieCellDelegate: MovieCellDelegate?
+    private var selectedMovieId: Int?
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Home/View/Reusable Views/MoviesTableViewCell.swift:22]: don't need this?

File: EnMovie/Modules/Home/View/Reusable Views/MoviesTableViewCell.swift [hunk 23]
```diff
@@ -35,6 +42,10 @@ class MoviesTableViewCell: UITableViewCell {
 
 //MARK: UICollectionViewDelegate
 extension MoviesTableViewCell: UICollectionViewDelegate {
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        selectedMovieId = upcomingMoviesArray[indexPath.row].id
+        self.movieCellDelegate?.cellWasPressed(movieId: selectedMovieId ?? 0)
```
Comment (iOSinternshipen) [inline EnMovie/Modules/Home/View/Reusable Views/MoviesTableViewCell.swift:47]: you can send directly upcomingMoviesArray[indexPath.row].id without and don't need to have a private var for it

**Review by iOSinternshipen — CHANGES_REQUESTED** [review]

**Review by barna-fulop — APPROVED** [review]

**Review by GeorgeCuciureanu — APPROVED** [review]

**Review by iulia-anisoi-endava — APPROVED** [review]

**Review by iOSinternshipen — APPROVED** [review]

**Pull request merged**